File: apps/mobile/ios/Fabric/Core/JsonRpcGatewayClient.swift
import Foundation

/// Connection lifecycle, mirroring `ConnectionState` in
/// `apps/shared/src/json-rpc-gateway.ts`.
enum GatewayConnectionState: String {
    case idle
    case connecting
    case open
    case closed
    case error
}

/// A server-pushed frame: `{"jsonrpc":"2.0","method":"event","params":{...}}`.
/// `type` values are the `GatewayEventName` strings shared with the desktop
/// renderer (`message.delta`, `tool.start`, `approval.request`, ...).
struct GatewayEvent {
    let type: String
    let sessionId: String?
    let payload: [String: Any]

    var payloadText: String? { payload["text"] as? String }
}

enum GatewayClientError: LocalizedError {
    case notConnected
    case connectFailed(underlying: String?)
    case socketClosed
    case requestTimedOut(method: String)
    case rpc(message: String, code: Int? = nil, data: Any? = nil)

    var errorDescription: String? {
        switch self {
        case .notConnected:
            return "gateway not connected"
        case .connectFailed(let underlying):
            let suffix = underlying.map { ": \($0)" } ?? ""
            return "WebSocket connection failed\(suffix)"
        case .socketClosed:
            return "WebSocket closed"
        case .requestTimedOut(let method):
            return "request timed out: \(method)"
        case .rpc(let message, _, _):
            return message
        }
    }

    static func rpc(body: [String: Any]) -> GatewayClientError {
        .rpc(
            message: body["message"] as? String ?? "Fabric RPC failed",
            code: body["code"] as? Int,
            data: body["data"]
        )
    }
}

/// JSON-RPC 2.0 client over a single WebSocket.
///
/// Swift port of `apps/shared/src/json-rpc-gateway.ts` — same wire contract
/// the desktop renderer uses against `fabric serve` (`/api/ws`): string ids
/// with an `"r"` prefix, a pending-request map with per-request timeouts,
/// response frames keyed by id, and unsolicited `method == "event"` frames
/// fanned out to subscribers.
///
/// All mutable state is confined to `stateQueue`. Event and state callbacks
/// are delivered on the main queue so SwiftUI observers can consume them
/// directly.
final class JsonRpcGatewayClient: NSObject {
    static let defaultRequestTimeout: TimeInterval = 120
    static let defaultConnectTimeout: TimeInterval = 15

    /// Foundation defaults WebSocket messages to 1 MiB. Some expected Fabric
    /// responses (notably an enabled pet spritesheet) legitimately exceed that
    /// and otherwise make iOS close an already-authenticated socket with 1009.
    /// Keep explicit headroom without accepting unbounded gateway frames.
    private static let maximumInboundMessageSize = 8 * 1024 * 1024

    var onEvent: ((GatewayEvent) -> Void)?
    var onStateChange: ((GatewayConnectionState) -> Void)?

    private let stateQueue = DispatchQueue(label: "io.github.obliviousodin.fabric.gateway-client")
    private lazy var urlSession = URLSession(
        configuration: .default,
        delegate: self,
        delegateQueue: nil
    )

    private var socket: URLSessionWebSocketTask?
    private var socketURL: URL?
    private var state: GatewayConnectionState = .idle
    private var nextId = 0
    private var pendingRequests: [String: (Result<Any?, Error>) -> Void] = [:]
    private var pendingTimeouts: [String: DispatchWorkItem] = [:]
    private var connectAttemptId = 0
    private var connectContinuation: (id: Int, continuation: CheckedContinuation<Void, Error>)?
    private var connectTimeoutItem: DispatchWorkItem?

    var connectionState: GatewayConnectionState {
        stateQueue.sync { state }
    }

    // MARK: - Connect / close

    func connect(to wsURL: URL, timeout: TimeInterval = JsonRpcGatewayClient.defaultConnectTimeout) async throws {
        let attemptId: Int? = stateQueue.sync {
            if let socket,
               socket.state == .running,
               state == .open,
               socketURL == wsURL {
                return nil
            }

            connectAttemptId += 1
            let attemptId = connectAttemptId
            connectTimeoutItem?.cancel()
            connectTimeoutItem = nil
            let previousConnect = connectContinuation
            connectContinuation = nil
            socket?.cancel()
            socket = nil
            socketURL = nil
            previousConnect?.continuation.resume(
                throwing: GatewayClientError.connectFailed(
                    underlying: "superseded by a newer connection"
                )
            )
            rejectAllPendingLocked(with: GatewayClientError.socketClosed)
            setStateLocked(.connecting)
            return attemptId
        }
        guard let attemptId else { return }

        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
            stateQueue.async {
                guard self.state == .connecting,
                      self.connectAttemptId == attemptId else {
                    continuation.resume(throwing: GatewayClientError.socketClosed)
                    return
                }
                let task = self.urlSession.webSocketTask(with: wsURL)
                Self.configureWebSocket(task)
                self.socket = task
                self.socketURL = wsURL
                self.connectContinuation = (attemptId, continuation)

                // A reconnect after sleep/wake must not hang in `.connecting`
                // forever (same rationale as DEFAULT_CONNECT_TIMEOUT_MS in the
                // shared TS client): fail to `.error` so callers can retry.
                let timeoutItem = DispatchWorkItem { [weak self] in
                    guard let self else { return }
                    guard self.connectAttemptId == attemptId,
                          self.socket === task,
                          let pending = self.connectContinuation,
                          pending.id == attemptId else { return }
                    self.connectContinuation = nil
                    self.connectTimeoutItem = nil
                    task.cancel()
                    self.socket = nil
                    self.socketURL = nil
                    self.setStateLocked(.error)
                    pending.continuation.resume(
                        throwing: GatewayClientError.connectFailed(underlying: "timed out")
                    )
                }
                self.connectTimeoutItem = timeoutItem
                self.stateQueue.asyncAfter(deadline: .now() + timeout, execute: timeoutItem)

                task.resume()
            }
        }
    }

    static func configureWebSocket(_ task: URLSessionWebSocketTask) {
        task.maximumMessageSize = maximumInboundMessageSize
    }

    func close() {
        stateQueue.async {
            self.connectAttemptId += 1
            self.connectTimeoutItem?.cancel()
            self.connectTimeoutItem = nil
            let pendingConnect = self.connectContinuation
            self.connectContinuation = nil
            self.socket?.cancel(with: .normalClosure, reason: nil)
            self.socket = nil
            self.socketURL = nil
            self.setStateLocked(.closed)
            pendingConnect?.continuation.resume(throwing: GatewayClientError.socketClosed)
            self.rejectAllPendingLocked(with: GatewayClientError.socketClosed)
        }
    }

    // MARK: - Requests

    /// Send a JSON-RPC request and await its response frame. `params` must be
    /// JSON-encodable. Returns the raw `result` value (dictionary for every
    /// Fabric method used here).
    func request(
        _ method: String,
        params: [String: Any] = [:],
        timeout: TimeInterval = JsonRpcGatewayClient.defaultRequestTimeout
    ) async throws -> Any? {
        try await withCheckedThrowingContinuation { continuation in
            stateQueue.async {
                guard let socket = self.socket, self.state == .open else {
                    continuation.resume(throwing: GatewayClientError.notConnected)
                    return
                }

                self.nextId += 1
                let id = "r\(self.nextId)"
                let frame: [String: Any] = [
                    "jsonrpc": "2.0",
                    "id": id,
                    "method": method,
                    "params": params,
                ]

                let data: Data
                do {
                    data = try JSONSerialization.data(withJSONObject: frame)
                } catch {
                    continuation.resume(throwing: error)
                    return
                }

                self.pendingRequests[id] = { result in
                    continuation.resume(with: result)
                }

                if timeout > 0 {
                    let timeoutItem = DispatchWorkItem { [weak self] in
                        guard let self else { return }
                        if let pending = self.pendingRequests.removeValue(forKey: id) {
                            self.pendingTimeouts.removeValue(forKey: id)
                            pending(.failure(GatewayClientError.requestTimedOut(method: method)))
                        }
                    }
                    self.pendingTimeouts[id] = timeoutItem
                    self.stateQueue.asyncAfter(deadline: .now() + timeout, execute: timeoutItem)
                }

                socket.send(.string(String(decoding: data, as: UTF8.self))) { [weak self] error in
                    guard let error, let self else { return }
                    self.stateQueue.async {
                        self.pendingTimeouts.removeValue(forKey: id)?.cancel()
                        if let pending = self.pendingRequests.removeValue(forKey: id) {
                            pending(.failure(error))
                        }
                    }
                }
            }
        }
    }

    /// Typed convenience: request expecting a JSON-object result.
    func requestObject(
        _ method: String,
        params: [String: Any] = [:],
        timeout: TimeInterval = JsonRpcGatewayClient.defaultRequestTimeout
    ) async throws -> [String: Any] {
        let raw = try await request(method, params: params, timeout: timeout)
        return raw as? [String: Any] ?? [:]
    }

    // MARK: - Receive loop

    private func startReceiveLoop(for socket: URLSessionWebSocketTask) {
        socket.receive { [weak self] result in
            guard let self else { return }
            self.stateQueue.async {
                guard self.socket === socket else { return }
                switch result {
                case .failure:
                    // The delegate close/error callbacks own state transitions;
                    // stopping the loop here is enough.
                    break
                case .success(let message):
                    switch message {
                    case .string(let text):
                        self.handleMessageLocked(text)
                    case .data(let data):
                        self.handleMessageLocked(String(decoding: data, as: UTF8.self))
                    @unknown default:
                        break
                    }
                    guard self.socket === socket else { return }
                    self.startReceiveLoop(for: socket)
                }
            }
        }
    }

    /// Parses a frame only after the receive loop has proved it belongs to the
    /// current socket. Must run on `stateQueue`.
    private func handleMessageLocked(_ text: String) {
        guard
            let data = text.data(using: .utf8),
            let frame = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return }

        // Response frame: routed to the matching pending request.
        if let id = frame["id"] as? String {
            pendingTimeouts.removeValue(forKey: id)?.cancel()
            guard let pending = pendingRequests.removeValue(forKey: id) else { return }
            if let errorBody = frame["error"] as? [String: Any] {
                pending(.failure(GatewayClientError.rpc(body: errorBody)))
            } else {
                pending(.success(frame["result"]))
            }
            return
        }

        // Event frame.
        guard
            frame["method"] as? String == "event",
            let params = frame["params"] as? [String: Any],
            let type = params["type"] as? String
        else { return }

        let event = GatewayEvent(
            type: type,
            sessionId: params["session_id"] as? String,
            payload: params["payload"] as? [String: Any] ?? [:]
        )
        let handler = onEvent
        DispatchQueue.main.async { handler?(event) }
    }

    // MARK: - State plumbing

    /// Must run on `stateQueue`.
    private func setStateLocked(_ newState: GatewayConnectionState) {
        guard state != newState else { return }
        state = newState
        let handler = onStateChange
        DispatchQueue.main.async { handler?(newState) }
    }

    /// Must run on `stateQueue`.
    private func rejectAllPendingLocked(with error: Error) {
        for (_, item) in pendingTimeouts { item.cancel() }
        pendingTimeouts.removeAll()
        let callbacks = pendingRequests.values
        pendingRequests.removeAll()
        for callback in callbacks {
            callback(.failure(error))
        }
    }
}

// MARK: - URLSessionWebSocketDelegate

extension JsonRpcGatewayClient: URLSessionWebSocketDelegate {
    func urlSession(
        _ session: URLSession,
        webSocketTask: URLSessionWebSocketTask,
        didOpenWithProtocol protocol: String?
    ) {
        stateQueue.async {
            guard self.socket === webSocketTask else { return }
            self.connectTimeoutItem?.cancel()
            self.connectTimeoutItem = nil
            self.setStateLocked(.open)
            self.startReceiveLoop(for: webSocketTask)
            self.connectContinuation?.continuation.resume()
            self.connectContinuation = nil
        }
    }

    func urlSession(
        _ session: URLSession,
        webSocketTask: URLSessionWebSocketTask,
        didCloseWith closeCode: URLSessionWebSocketTask.CloseCode,
        reason: Data?
    ) {
        stateQueue.async {
            guard self.socket === webSocketTask else { return }
            self.socket = nil
            self.socketURL = nil
            self.connectTimeoutItem?.cancel()
            self.connectTimeoutItem = nil
            if let pending = self.connectContinuation {
                self.connectContinuation = nil
                self.setStateLocked(.error)
                pending.continuation.resume(
                    throwing: GatewayClientError.connectFailed(
                        underlying: "closed (code \(closeCode.rawValue))"
                    )
                )
            } else {
                self.setStateLocked(.closed)
            }
            self.rejectAllPendingLocked(with: GatewayClientError.socketClosed)
        }
    }

    func urlSession(
        _ session: URLSession,
        task: URLSessionTask,
        didCompleteWithError error: Error?
    ) {
        guard let error else { return }
        stateQueue.async {
            guard self.socket === task else { return }
            self.socket = nil
            self.socketURL = nil
            self.connectTimeoutItem?.cancel()
            self.connectTimeoutItem = nil
            if let pending = self.connectContinuation {
                self.connectContinuation = nil
                self.setStateLocked(.error)
                pending.continuation.resume(
                    throwing: GatewayClientError.connectFailed(underlying: error.localizedDescription)
                )
            } else {
                self.setStateLocked(.error)
            }
            self.rejectAllPendingLocked(with: error)
        }
    }
}

File: apps/mobile/android/app/src/main/kotlin/io/github/obliviousodin/fabric/mobile/core/JsonRpcGatewayClient.kt
package io.github.obliviousodin.fabric.mobile.core

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withTimeout
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.put
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString

/** Connection lifecycle, mirroring `ConnectionState` in apps/shared/src/json-rpc-gateway.ts. */
enum class GatewayConnectionState { IDLE, CONNECTING, OPEN, CLOSED, ERROR }

/**
 * A server-pushed frame: `{"jsonrpc":"2.0","method":"event","params":{...}}`.
 * `type` values are the GatewayEventName strings shared with the desktop
 * renderer (message.delta, tool.start, approval.request, ...).
 */
data class GatewayEvent(
    val type: String,
    val sessionId: String?,
    val payload: JsonObject,
) {
    val payloadText: String?
        get() = (payload["text"] as? kotlinx.serialization.json.JsonPrimitive)?.contentOrNullSafe()
}

private fun kotlinx.serialization.json.JsonPrimitive.contentOrNullSafe(): String? =
    if (this is kotlinx.serialization.json.JsonNull) null else content

class GatewayRpcException(
    message: String,
    val code: Int? = null,
    val data: JsonElement? = null,
) : Exception(message)

internal fun gatewayRpcException(error: JsonObject): GatewayRpcException {
    val message = (error["message"] as? kotlinx.serialization.json.JsonPrimitive)
        ?.contentOrNullSafe() ?: "Fabric RPC failed"
    val code = (error["code"] as? kotlinx.serialization.json.JsonPrimitive)?.content?.toIntOrNull()
    return GatewayRpcException(message, code, error["data"])
}

class GatewayNotConnectedException : Exception("gateway not connected")

class GatewayConnectException(message: String, cause: Throwable? = null) : Exception(message, cause)

/** An RPC may have reached the gateway before its response timed out. */
class GatewayRequestTimeoutException(method: String) : Exception("request timed out: $method")

/**
 * A WebSocket failed after a request was handed to OkHttp but before its
 * response arrived. Callers must treat a mutation's outcome as unknown and
 * replay it only with the original idempotency key.
 */
class GatewayResponseUncertainException(message: String, cause: Throwable? = null) :
    Exception(message, cause)

/**
 * JSON-RPC 2.0 client over a single WebSocket.
 *
 * Kotlin port of `apps/shared/src/json-rpc-gateway.ts` — the same wire
 * contract the desktop renderer uses against `fabric serve` (`/api/ws`):
 * string ids with an `"r"` prefix, a pending-request map with per-request
 * timeouts, response frames keyed by id, and unsolicited `method == "event"`
 * frames fanned out to subscribers.
 */
class JsonRpcGatewayClient(
    httpClient: OkHttpClient? = null,
) {
    companion object {
        const val DEFAULT_REQUEST_TIMEOUT_MS = 120_000L

        // A reconnect after sleep/wake must not hang in CONNECTING forever;
        // fail to ERROR so callers can retry (same rationale as the TS client).
        const val DEFAULT_CONNECT_TIMEOUT_MS = 15_000L
    }

    private val json = Json { ignoreUnknownKeys = true }

    private val client: OkHttpClient = httpClient
        ?: OkHttpClient.Builder()
            // JSON-RPC requests carry their own per-call timeouts.
            .readTimeout(0, TimeUnit.MILLISECONDS)
            .pingInterval(30, TimeUnit.SECONDS)
            .build()

    private val nextId = AtomicInteger(0)
    private val pending = ConcurrentHashMap<String, CompletableDeferred<JsonElement?>>()

    @Volatile
    private var socket: WebSocket? = null

    @Volatile
    private var socketUrl: String? = null

    @Volatile
    private var connectHandshake: CompletableDeferred<Unit>? = null

    private val connectionLock = Any()

    private val _state = MutableStateFlow(GatewayConnectionState.IDLE)
    val state: StateFlow<GatewayConnectionState> = _state.asStateFlow()

    private val _events = MutableSharedFlow<GatewayEvent>(extraBufferCapacity = 256)
    val events: SharedFlow<GatewayEvent> = _events.asSharedFlow()

    suspend fun connect(wsUrl: String, timeoutMs: Long = DEFAULT_CONNECT_TIMEOUT_MS) {
        lateinit var handshake: CompletableDeferred<Unit>
        lateinit var attemptSocket: WebSocket
        synchronized(connectionLock) {
            if (_state.value == GatewayConnectionState.OPEN &&
                socket != null &&
                socketUrl == wsUrl
            ) return

            connectHandshake?.completeExceptionally(
                GatewayConnectException("WebSocket connection superseded")
            )
            socket?.cancel()
            rejectAllPending(GatewayNotConnectedException())
            _state.value = GatewayConnectionState.CONNECTING
            handshake = CompletableDeferred()
            connectHandshake = handshake
            socketUrl = wsUrl

            val request = Request.Builder().url(wsUrl).build()
            attemptSocket = client.newWebSocket(request, listener)
            socket = attemptSocket
        }

        try {
            withTimeout(timeoutMs) { handshake.await() }
        } catch (e: TimeoutCancellationException) {
            // Drop the half-open socket so the next connect() starts clean.
            synchronized(connectionLock) {
                if (socket === attemptSocket) {
                    attemptSocket.cancel()
                    socket = null
                    socketUrl = null
                    _state.value = GatewayConnectionState.ERROR
                }
            }
            throw GatewayConnectException("WebSocket connection timed out", e)
        } catch (e: CancellationException) {
            synchronized(connectionLock) {
                if (socket === attemptSocket) {
                    attemptSocket.cancel()
                    socket = null
                    socketUrl = null
                    _state.value = GatewayConnectionState.CLOSED
                }
            }
            throw e
        } finally {
            synchronized(connectionLock) {
                if (connectHandshake === handshake) connectHandshake = null
            }
        }
    }

    fun close() {
        synchronized(connectionLock) {
            val current = socket
            socket = null
            socketUrl = null
            val handshake = connectHandshake
            connectHandshake = null
            current?.cancel()
            _state.value = GatewayConnectionState.CLOSED
            handshake?.completeExceptionally(GatewayNotConnectedException())
            rejectAllPending(GatewayNotConnectedException())
        }
    }

    /**
     * Send a JSON-RPC request and await its response frame. Returns the raw
     * `result` element (a JsonObject for every Fabric method used here).
     */
    suspend fun request(
        method: String,
        params: JsonObject = buildJsonObject {},
        timeoutMs: Long = DEFAULT_REQUEST_TIMEOUT_MS,
    ): JsonElement? {
        val currentSocket = socket
        if (currentSocket == null || _state.value != GatewayConnectionState.OPEN) {
            throw GatewayNotConnectedException()
        }

        val id = "r${nextId.incrementAndGet()}"
        val frame = buildJsonObject {
            put("jsonrpc", "2.0")
            put("id", id)
            put("method", method)
            put("params", params)
        }

        val deferred = CompletableDeferred<JsonElement?>()
        pending[id] = deferred

        if (!currentSocket.send(frame.toString())) {
            pending.remove(id)
            throw GatewayNotConnectedException()
        }

        return try {
            if (timeoutMs > 0) {
                withTimeout(timeoutMs) { deferred.await() }
            } else {
                deferred.await()
            }
        } catch (e: TimeoutCancellationException) {
            throw GatewayRequestTimeoutException(method)
        } finally {
            pending.remove(id)
        }
    }

    suspend fun requestObject(
        method: String,
        params: JsonObject = buildJsonObject {},
        timeoutMs: Long = DEFAULT_REQUEST_TIMEOUT_MS,
    ): JsonObject = request(method, params, timeoutMs) as? JsonObject ?: buildJsonObject {}

    private val listener = object : WebSocketListener() {
        override fun onOpen(webSocket: WebSocket, response: Response) {
            synchronized(connectionLock) {
                if (webSocket !== socket) return
                _state.value = GatewayConnectionState.OPEN
                connectHandshake?.complete(Unit)
            }
        }

        override fun onMessage(webSocket: WebSocket, text: String) {
            if (webSocket !== socket) return
            handleFrame(text)
        }

        override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
            if (webSocket !== socket) return
            handleFrame(bytes.utf8())
        }

        override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
            synchronized(connectionLock) {
                if (webSocket !== socket) return
                socket = null
                socketUrl = null
                _state.value = GatewayConnectionState.CLOSED
                connectHandshake?.completeExceptionally(
                    GatewayConnectException("WebSocket closed during connect (code $code)")
                )
                rejectAllPending(GatewayNotConnectedException())
            }
        }

        override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
            synchronized(connectionLock) {
                if (webSocket !== socket) return
                socket = null
                socketUrl = null
                _state.value = GatewayConnectionState.ERROR
                connectHandshake?.completeExceptionally(
                    GatewayConnectException("WebSocket connection failed", t)
                )
                rejectAllPending(
                    GatewayResponseUncertainException(
                        "WebSocket failed while awaiting a response",
                        t,
                    )
                )
            }
        }
    }

    private fun handleFrame(text: String) {
        val frame = runCatching { json.parseToJsonElement(text).jsonObject }.getOrNull() ?: return

        // Response frame: routed to the matching pending request.
        val id = frame["id"]?.let { el ->
            (el as? kotlinx.serialization.json.JsonPrimitive)?.contentOrNullSafe()
        }
        if (id != null) {
            val call = pending.remove(id) ?: return
            val error = frame["error"] as? JsonObject
            if (error != null) {
                call.completeExceptionally(gatewayRpcException(error))
            } else {
                call.complete(frame["result"])
            }
            return
        }

        // Event frame.
        val method = (frame["method"] as? kotlinx.serialization.json.JsonPrimitive)?.contentOrNullSafe()
        if (method != "event") return
        val params = frame["params"] as? JsonObject ?: return
        val type = (params["type"] as? kotlinx.serialization.json.JsonPrimitive)?.contentOrNullSafe() ?: return

        _events.tryEmit(
            GatewayEvent(
                type = type,
                sessionId = (params["session_id"] as? kotlinx.serialization.json.JsonPrimitive)
                    ?.contentOrNullSafe(),
                payload = params["payload"] as? JsonObject ?: buildJsonObject {},
            )
        )
    }

    private fun rejectAllPending(cause: Throwable) {
        val calls = pending.values.toList()
        pending.clear()
        calls.forEach { it.completeExceptionally(cause) }
    }
}
